src/main/java/de/hysky/skyblocker/skyblock/CompactDamage.java

```diff
@@ -86,15 +88,99 @@ public static void compactDamage(ArmorStand entity) {
-		if (damage < 1_000_000) return format(damage / 1_000.0) + "k";
-		if (damage < 1_000_000_000) return format(damage / 1_000_000.0) + "M";
-		if (damage < 1_000_000_000_000L) return format(damage / 1_000_000_000.0) + "B";
-		return format(damage / 1_000_000_000_000.0) + "T"; //This will probably never be reached
+	/// We want precision to signify the *number of significant digits*, not the number of digits after the decimal.
+	/// For example:
+	/// 123,456,789 (precision 3) -> 123M
+	/// 12,345 (precision 4) -> 1.234k
+	private static String prettifyDamageNumber(final int damage, final int maxPrecision) {
+		int targetDamage = damage;
+		int targetPrecision = maxPrecision;
+		// First, round `damage` to `precision` places
+		// Otherwise inputs like `999,999, 3` will display as `1000k` instead of `1.00m`
+
+		int usedPrecision = baseTenDigits(targetDamage);
+		if (usedPrecision > targetPrecision) {
+			int powerToRoundTo = powersOfTen[usedPrecision - maxPrecision];
+			targetDamage = (int) (Math.round((double) targetDamage / powerToRoundTo) * powerToRoundTo);
+		} else if (targetPrecision > usedPrecision) {
+			// We don't want to ever display more decimal points than needed. For example,
+			// 999_999 with precision 7 should still display as 999.999k, not 999.9990k
+			targetPrecision = usedPrecision;
+		}
+
+		if (targetDamage < 1_000) return String.valueOf(targetDamage);
+		if (targetDamage < 1_000_000) return formatToPrecision(targetDamage / 1_000.0, targetPrecision) + "k";
+		if (targetDamage < 1_000_000_000) return formatToPrecision(targetDamage / 1_000_000.0, targetPrecision) + "m";
+		return formatToPrecision(targetDamage / 1_000_000_000.0, targetPrecision) + "b";
+	}
+
+	private static String formatToPrecision(double number, int precision) {
+		int usedPrecision = baseTenDigits((int) number);
+		int remainingPrecision = precision - usedPrecision;
+		if (remainingPrecision <= 0) {
+			int powerToRoundTo = powersOfTen[usedPrecision - precision];
+			return String.valueOf((int) (Math.round(number / powerToRoundTo) * powerToRoundTo));
+		}
+		return ("%." + remainingPrecision + "f").formatted(number);
+	}
+
+	private static int baseTwoDigits(int x) {
+		return 32 - Integer.numberOfLeadingZeros(x);
 	}
 
-	private static String format(double number) {
-		return ("%." + SkyblockerConfigManager.get().uiAndVisuals.compactDamage.precision + "f").formatted(number);
+	private static final int[] guesses = new int[]{
+			0, 0, 0, 0, 1, 1, 1, 2, 2, 2,
+			3, 3, 3, 3, 4, 4, 4, 5, 5, 5,
+			6, 6, 6, 6, 7, 7, 7, 8, 8, 8,
+			9, 9, 9
+	};
+
+	private static final int[] powersOfTen = new int[]{
+			1, 10, 100, 1000, 10000, 100000,
+			1000000, 10000000, 100000000, 1000000000,
+	};
+
+	/// Equivalent to floor(log10(x)) + 1
+	/// https://stackoverflow.com/a/25934909
+	private static int baseTenDigits(int x) {
+		int guess = guesses[baseTwoDigits(x)];
+		return guess + ((x >= powersOfTen[guess]) ? 1 : 0);
+	}
+
+	public static void main(String[] args) {
```

Comment:
Was this meant to be here. Please put them in the test file if it was meant to stay in the mod. You can make private methods protected or default and annotate `VisibleForTesting`.